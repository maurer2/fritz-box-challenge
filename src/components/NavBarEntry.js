import React, { Component } from 'react';
import styled from 'styled-components/macro';
import PropTypes from 'prop-types';
import upperFirst from 'lodash/upperFirst';

const NavBarEntryWrapper = styled.li`
  flex-grow: 1;
  flex-shrink: 0;
  flex-basis: 0;
  background: ${props => (props.isActive ? '#cc0000' : 'transparent')};
`;

const defaultButton = styled.button`
  padding: 1rem;
  appearance: none;
  background: none;
  outline: none;
`;

const NavBarButton = styled(defaultButton)`
  display: block;
  width: 100%;
  border: 0;
  font-size: 1rem;
`;

class NavBarEntry extends Component {
  constructor(props) {
    super(props);

    this.elementRef = React.createRef();
    this.handleClick = this.handleClick.bind(this);
  }

  getBoundingBox() {
    if (this.elementRef.current === null) {
      return 0;
    }

    const element = this.elementRef.current;

    return element.getBoundingClientRect();
  }

  componentDidMount() {
    const { isActive } = this.props;

    if (isActive) {
      const elementBoundingBox = this.getBoundingBox();

      this.props.handleIndicatorUpdate(elementBoundingBox.x, elementBoundingBox.width);
    }
  }

  componentDidUpdate(prevProps) {
    const isNewActive = (this.props.isActive && !(prevProps.isActive));

    if (isNewActive) {
      const elementBoundingBox = this.getBoundingBox();

      this.props.handleIndicatorUpdate(elementBoundingBox.x, elementBoundingBox.width);
    }
  }

  handleClick() {
    this.props.handleNavigation(this.props.index);
  }

  render() {
    const { isActive, entry } = this.props;

    return (
      <NavBarEntryWrapper isActive={ isActive } onClick={ this.handleClick } ref={ this.elementRef }>
        <NavBarButton>
          { upperFirst(entry) }
        </NavBarButton>
      </NavBarEntryWrapper>
    );
  }
}

NavBarEntry.propTypes = {
  index: PropTypes.number,
  entry: PropTypes.string,
  isActive: PropTypes.bool,
  handleNavigation: PropTypes.func,
  handleIndicatorUpdate: PropTypes.func,
};

export default NavBarEntry;
