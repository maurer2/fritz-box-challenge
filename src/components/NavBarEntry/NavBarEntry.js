import React from 'react';
import styled from 'styled-components/macro';
import PropTypes from 'prop-types';
import upperFirst from 'lodash/upperFirst';

const NavBarEntryWrapper = styled.li`
  flex-grow: 1;
  flex-shrink: 0;
  flex-basis: ${props => (props.isFullWidth ? '33%' : '0')};
`;

const defaultButton = styled.button`
  padding: 1rem;
  appearance: none;
  background: none;
  outline: none;
`;

const NavBarButton = styled(defaultButton)`
  display: block;
  width: 100%;
  border: 0;
  font-size: 1rem;
  color: ${props => (props.isActive ? '#080808' : '#121212')};
  font-weight: ${props => (props.isActive ? 'bold' : 'normal')};
  transition: color 500ms;
`;

const NavBarEntry = ({ index, entry, isActive, handleNavigation, activeElementRef, isFullWidth }) => {
  function handleClick() {
    handleNavigation(index);
  }

  return (
    <NavBarEntryWrapper
      onClick={ handleClick }
      isFullWidth={ isFullWidth }
      ref={ activeElementRef }
    >
      <NavBarButton isActive={ isActive }>
        { upperFirst(entry) }
      </NavBarButton>
    </NavBarEntryWrapper>
  );
};

NavBarEntry.propTypes = {
  index: PropTypes.number,
  entry: PropTypes.string,
  isActive: PropTypes.bool,
  handleNavigation: PropTypes.func,
  activeElementRef: PropTypes.any,
  isFullWidth: PropTypes.bool,
};

export { NavBarEntry };
