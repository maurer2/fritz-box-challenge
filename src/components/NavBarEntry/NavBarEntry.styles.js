import styled from 'styled-components/macro';

export const NavBarEntryWrapper = styled.li`
  flex: 1 0 ${(props) => (props.isFullWidth ? '33%' : '0')};
`;

export const defaultButton = styled.button`
  padding: 1rem;
  appearance: none;
  background: none;
  outline: none;
`;

export const NavBarButton = styled(defaultButton)`
  display: block;
  width: 100%;
  border: 0;
  font-size: 1rem;
  color: ${({ theme }) => theme.tertiaryColor};
  font-weight: ${({ isActive }) => (isActive ? 'bold' : 'normal')};
  transition: color 500ms;
`;
