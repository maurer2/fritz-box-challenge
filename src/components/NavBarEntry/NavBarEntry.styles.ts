import styled from 'styled-components/macro';

import type { NavBarEntryWrapperStyleProps, NavBarButtonStyleProps } from './NavBarEntry.types';

export const NavBarEntryWrapper = styled.li<NavBarEntryWrapperStyleProps>`
  flex: 1 0 ${(props) => (props.isFullWidth ? '33%' : '0')};
`;

export const defaultButton = styled.button`
  padding: 1rem;
  background: none;
  outline: none;
  appearance: none;
`;

export const NavBarButton = styled(defaultButton)<NavBarButtonStyleProps>`
  display: block;
  width: 100%;
  border: 0;
  color: ${({ theme }) => theme.tertiaryColor};
  font-weight: ${({ isActive }) => (isActive ? 'bold' : 'normal')};
  font-size: 1rem;
  transition: color 500ms;
`;
