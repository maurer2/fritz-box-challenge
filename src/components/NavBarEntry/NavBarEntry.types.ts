export interface NavBarEntryProps {
  index: number;
  entry: string;
  $isActive: boolean;
  handleNavigation: any;
  activeElementRef: React.RefObject<HTMLLIElement>;
  $isFullWidth: boolean;
}

export interface NavBarEntryWrapperStyleProps {
  $isFullWidth: boolean;
}

export interface NavBarButtonStyleProps {
  $isActive: boolean;
}
