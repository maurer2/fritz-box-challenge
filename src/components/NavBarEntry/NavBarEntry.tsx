import React from 'react';
import PropTypes from 'prop-types';
import { upperFirst } from 'lodash';

import * as Styles from './NavBarEntry.styles';
import * as Types from './NavBarEntry.types';

const NavBarEntry: React.FC<Types.NavBarEntryProps> = ({ index,
  entry,
  isActive,
  handleNavigation,
  activeElementRef,
  isFullWidth }): JSX.Element => {
  function handleClick(): void {
    handleNavigation(index);
  }

  return (
    <Styles.NavBarEntryWrapper
      onClick={handleClick}
      isFullWidth={isFullWidth}
      ref={activeElementRef}
    >
      <Styles.NavBarButton isActive={isActive}>{upperFirst(entry)}</Styles.NavBarButton>
    </Styles.NavBarEntryWrapper>
  );
};

const { string, number, bool, func, shape, instanceOf } = PropTypes;

NavBarEntry.defaultProps = {
  activeElementRef: shape({
    current: null,
  }),
};

NavBarEntry.propTypes = {
  index: number.isRequired,
  entry: string.isRequired,
  isActive: bool.isRequired,
  handleNavigation: func.isRequired,
  activeElementRef: shape({
    current: instanceOf(HTMLElement),
  }),
  isFullWidth: bool.isRequired,
};

export { NavBarEntry };
