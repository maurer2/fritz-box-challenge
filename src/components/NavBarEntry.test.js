import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import NavBarEntry from './NavBarEntry';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

describe('NavBarEntry', () => {
  const wrapper = shallow(<NavBarEntry entry="model" />);

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a title component', () => {
    expect(wrapper.find('NavBarEntryWrapper').length).toBe(1);
  });

  it('should have text in component', () => {
    expect(wrapper.find('NavBarEntryWrapper').text()).toBe('Model');
  });
});
