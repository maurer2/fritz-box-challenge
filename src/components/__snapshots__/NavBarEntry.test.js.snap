// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBarEntry should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NavBarEntry
    entry="model"
    handleNavigation={[MockFunction]}
    index={5}
    isActive={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <ForwardRef>
        Model
      </ForwardRef>,
      "isActive": false,
      "onClick": [Function],
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": "Model",
      },
      "ref": null,
      "rendered": "Model",
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "NavBarButton-sc-1jim1v6-2",
          "isStatic": true,
          "rules": Array [
            "padding:1rem;appearance:none;background:none;outline:none;",
            "display:block;width:100%;border:0;font-size:1rem;",
          ],
        },
        "displayName": "NavBarButton",
        "foldedComponentIds": Array [
          "defaultButton-sc-1jim1v6-1",
        ],
        "render": [Function],
        "styledComponentId": "NavBarButton-sc-1jim1v6-2",
        "target": "button",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "NavBarEntryWrapper-sc-1jim1v6-0",
        "isStatic": false,
        "rules": Array [
          "flex-grow:1;flex-shrink:0;flex-basis:0;background:",
          [Function],
          ";",
        ],
      },
      "displayName": "NavBarEntryWrapper",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "NavBarEntryWrapper-sc-1jim1v6-0",
      "target": "li",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <ForwardRef>
          Model
        </ForwardRef>,
        "isActive": false,
        "onClick": [Function],
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Model",
        },
        "ref": null,
        "rendered": "Model",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "NavBarButton-sc-1jim1v6-2",
            "isStatic": true,
            "rules": Array [
              "padding:1rem;appearance:none;background:none;outline:none;",
              "display:block;width:100%;border:0;font-size:1rem;",
            ],
          },
          "displayName": "NavBarButton",
          "foldedComponentIds": Array [
            "defaultButton-sc-1jim1v6-1",
          ],
          "render": [Function],
          "styledComponentId": "NavBarButton-sc-1jim1v6-2",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "NavBarEntryWrapper-sc-1jim1v6-0",
          "isStatic": false,
          "rules": Array [
            "flex-grow:1;flex-shrink:0;flex-basis:0;background:",
            [Function],
            ";",
          ],
        },
        "displayName": "NavBarEntryWrapper",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "NavBarEntryWrapper-sc-1jim1v6-0",
        "target": "li",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
